### PR TITLE
more efficient atlas packing algorithm

### PR DESCRIPTION
**Pull request by DrMoriarty**
I changed the atlas packing algorithm.

Atlas before: 
atlas_before

Atlas after: 
atlas_after

We can reach even more efficiency if we able to rotate the images. But I don't know is it possible.

### CODE REVIEW

File: editor/io_plugins/editor_atlas.cpp [hunk 1]
```diff
@@ -45,23 +46,23 @@ struct _EditorAtlasWorkRectResult {
...
-	//it will work in every hardware.
-	// for example, it will prioritize a 1024x1024 atlas (works everywhere) instead of a
-	// 256x8192 atlas (won't work anywhere).
+void EditorAtlas::fit(const Vector<Size2i> &p_rects, Vector<Point2i> &r_result, Size2i &r_size) {
 
 	ERR_FAIL_COND(p_rects.size() == 0);
 
 	Vector<_EditorAtlasWorkRect> wrects;
 	wrects.resize(p_rects.size());
+    long Stotal = 0;
```
Comment (akien-mga) [inline editor/io_plugins/editor_atlas.cpp]: Indentation issue.

File: editor/io_plugins/editor_atlas.cpp [hunk 2]
```diff
@@ -76,68 +77,90 @@ void EditorAtlas::fit(const Vector<Size2i> &p_rects, Vector<Point2i> &r_result,
...
-			wrects[j].p.y = from_y;
-
-			int end_h = from_y + wrects[j].s.height;
-			int end_w = ofs + wrects[j].s.width;
-
-			for (int k = 0; k < wrects[j].s.width; k++) {
-
-				hmax[ofs + k] = end_h;
-			}
+            int newX = 0;
```
Comment (akien-mga) [inline editor/io_plugins/editor_atlas.cpp]: Please use snake_case for identifiers and methods.

File: editor/io_plugins/editor_atlas.cpp [hunk 3]
```diff
@@ -76,68 +77,90 @@ void EditorAtlas::fit(const Vector<Size2i> &p_rects, Vector<Point2i> &r_result,
...
-
-			ofs += wrects[j].s.width;
 		}
 
 		_EditorAtlasWorkRectResult result;
 		result.result = wrects;
 		result.max_h = max_h;
 		result.max_w = max_w;
 		results.push_back(result);
+        float efficiency = float(max_w * max_h) / float(nearestPow2(max_w) * nearestPow2(max_h));
```
Comment (akien-mga) [inline editor/io_plugins/editor_atlas.cpp]: Indentation issue.

File: editor/io_plugins/editor_atlas.cpp [hunk 4]
```diff
@@ -76,68 +77,90 @@ void EditorAtlas::fit(const Vector<Size2i> &p_rects, Vector<Point2i> &r_result,
...
-	float best_aspect = 1e20;
+	float max_eff = 0;
 
 	for (int i = 0; i < results.size(); i++) {
 
 		float h = results[i].max_h;
 		float w = results[i].max_w;
-		float aspect = h > w ? h / w : w / h;
-		if (aspect < best_aspect) {
+        float efficiency = float(w * h) / float(nearestPow2(w) * nearestPow2(h));
```
Comment (akien-mga) [inline editor/io_plugins/editor_atlas.cpp]: Indentation issue.

File: editor/io_plugins/editor_atlas.cpp [hunk 5]
```diff
@@ -29,6 +29,7 @@
 /*************************************************************************/
 #include "editor_atlas.h"
 #include "print_string.h"
+#include <cfloat>
```
Comment (akien-mga) [inline editor/io_plugins/editor_atlas.cpp]: Why is it needed?

File: editor/io_plugins/editor_atlas.cpp [hunk 6]
```diff
@@ -45,23 +46,23 @@ struct _EditorAtlasWorkRectResult {
 	int max_h;
 };
 
-void EditorAtlas::fit(const Vector<Size2i> &p_rects, Vector<Point2i> &r_result, Size2i &r_size) {
+static int nearestPow2(int d) {
```
Comment (akien-mga) [inline editor/io_plugins/editor_atlas.cpp]: You could use one of those instead: https://github.com/godotengine/godot/blob/1a92906b681735a5a2acfd98e3d6a4c75a7f6115/core/typedefs.h#L167-L194

File: editor/io_plugins/editor_atlas.cpp [hunk 7]
```diff
@@ -76,68 +77,90 @@ void EditorAtlas::fit(const Vector<Size2i> &p_rects, Vector<Point2i> &r_result,
...
+
+            int pieceW = wrects[j].s.width;
+            int pieceH = wrects[j].s.height;
+
+            bool foundPlace;
+
+            do {
+                foundPlace = true;
+                newX = 0;
+                if(wmax.size() <= newY + pieceH) {
```
Comment (akien-mga) [inline editor/io_plugins/editor_atlas.cpp]: There are some style issues, e.g. missing space between `if` and `(`. Check docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html for more info on the code style and how to apply clang-format locally.

**Comment by DrMoriarty** [comment]
@akien-mga fixed

**Comment by reduz** [comment]
This is great! though atlas is currently not used in 3.0.. I was thinking
of ways to re-implement it but i'm not completely sure how..

**Comment by akien-mga** [comment]
Looks good, but there are still some code style issues. Could you:
- Rebase your branch to squash the 4 commits together? See http://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase
- Ensure that you follow the clang-format-enforced code style; you can test locally with http://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html#using-clang-format-locally

**Comment by DrMoriarty** [comment]
@akien-mga done

**Comment by DrMoriarty** [comment]
@akien-mga also could you suggest me how can I implement atlases with rotated images?
I think this will be very useful.

**Comment by akien-mga** [comment]
> also could you suggest me how can I implement atlases with rotated images?

No idea, sorry :)

**Pull request merged**